nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/CaptureChangeMySQL.java

```diff
@@ -0,0 +1,928 @@
+        }
+
+        ComponentLog log = getLogger();
+        StateManager stateManager = context.getStateManager();
+
+        try {
+            outputEvents(currentSession, stateManager, log);
+        } catch (IOException ioe) {
+            try {
+                // Perform some processor-level "rollback", then rollback the session
+                currentBinlogFile = xactBinlogFile == null ? "" : xactBinlogFile;
+                currentBinlogPosition = xactBinlogPosition;
+                currentSequenceId.set(xactSequenceId);
+                inTransaction = false;
+                stop(stateManager);
+                queue.clear();
+                currentSession.rollback();
+            } catch (Exception e) {
+                // Not much we can recover from here
+            }
+            throw new ProcessException(ioe);
+        }
+    }
+
+    @OnStopped
+    public void onStopped(ProcessContext context) {
+        try {
+            stop(context.getStateManager());
+        } catch (CDCException ioe) {
+            throw new ProcessException(ioe);
+        }
+    }
+
+    /**
+     * Get a list of hosts from a NiFi property, e.g.
+     *
+     * @param hostsString A comma-separated list of hosts (host:port,host2:port2, etc.)
+     * @return List of InetSocketAddresses for the hosts
+     */
+    private List<InetSocketAddress> getHosts(String hostsString) {
+
+        if (hostsString == null) {
+            return null;
+        }
+        final List<String> hostsSplit = Arrays.asList(hostsString.split(","));
+        List<InetSocketAddress> hostsList = new ArrayList<>();
+
+        for (String item : hostsSplit) {
+            String[] addresses = item.split(":");
+            if (addresses.length != 2) {
+                throw new ArrayIndexOutOfBoundsException("Not in host:port format");
+            }
+
+            hostsList.add(new InetSocketAddress(addresses[0].trim(), Integer.parseInt(addresses[1].trim())));
+        }
+        return hostsList;
+    }
+
+    protected void connect(List<InetSocketAddress> hosts, String username, String password, boolean createEnrichmentConnection,
+                           String driverLocation, String driverName, long connectTimeout) throws IOException {
+
+        int connectionAttempts = 0;
+        final int numHosts = hosts.size();
+        InetSocketAddress connectedHost = null;
+
+        while (connectedHost == null && connectionAttempts < numHosts) {
+            if (binlogClient == null) {
+
+                connectedHost = hosts.get(currentHost);
+                binlogClient = createBinlogClient(connectedHost.getHostString(), connectedHost.getPort(), username, password);
+            }
+
+            if (eventListener == null) {
+                eventListener = createBinlogEventListener(binlogClient, queue);
+            }
+            eventListener.start();
+            binlogClient.registerEventListener(eventListener);
```

Comment:
Yeah I was curious about LifecycleListener myself but hadn't seen the need. This is obviously one of them :)  Will make the change.